### ISSUE TITLE
llvm-prepo: Incorrect hash calculation for the ‘CallInst’ instruction. 

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-prepo#33]
This bug is spinning out from bug 32.

Compile the following code:
```
target datalayout = "e-m:r-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-repo"

%struct._xmlParserCtxt = type opaque
%struct._xmlHashTable = type opaque
%struct._xmlEntity = type opaque

define void @htmlFreeParserCtxt(%struct._xmlParserCtxt* %ctxt) {
entry:
  %ctxt.addr = alloca %struct._xmlParserCtxt*, align 8
  store %struct._xmlParserCtxt* %ctxt, %struct._xmlParserCtxt** %ctxt.addr, align 8
  %0 = load %struct._xmlParserCtxt*, %struct._xmlParserCtxt** %ctxt.addr, align 8
  call void @xmlFreeParserCtxt(%struct._xmlParserCtxt* %0)
  ret void
}
declare void @xmlFreeParserCtxt(%struct._xmlParserCtxt*)

define void @xmlFreeEntitiesTable(%struct._xmlHashTable* %table) {
entry:
  %table.addr = alloca %struct._xmlHashTable*, align 8
  store %struct._xmlHashTable* %table, %struct._xmlHashTable** %table.addr, align 8
  %0 = load %struct._xmlHashTable*, %struct._xmlHashTable** %table.addr, align 8
  call void @xmlHashFree(%struct._xmlHashTable* %0, void (i8*, i8*)* bitcast (void (%struct._xmlEntity*, i8*)* @xmlFreeEntityWrapper to void (i8*, i8*)*))
  ret void
}
declare void @xmlHashFree(%struct._xmlHashTable*, void (i8*, i8*)*)
declare void @xmlFreeEntityWrapper(%struct._xmlEntity* %entity, i8* %name)
```
using
```
$ rm clang.db
$ opt test.ll -S -o  test.opt.ll
```
The generated optimised test.opt.ll file is:

```
define void @htmlFreeParserCtxt(%struct._xmlParserCtxt* %ctxt) !repo_ticket !0 {
...
}
define void @xmlFreeEntitiesTable(%struct._xmlHashTable* %table) !repo_ticket !1 {
...
}
!0 = !TicketNode(name: "htmlFreeParserCtxt", digest: [16 x i8] c"\CF\B3\F7h\E8\F9\E7\0C~&\92\F0\1AG\A2\E5", linkage: external, pruned: false)
!1 = !TicketNode(name: "xmlFreeEntitiesTable", digest: [16 x i8] c"\CF\B3\F7h\E8\F9\E7\0C~&\92\F0\1AG\A2\E5", linkage: external, pruned: false)
```
The functions of “htmlFreeParserCtxt” and “xmlFreeEntitiesTable” have the same digest, which is wrong. The digests should be different.

**Comment by MaggieYingYi** [comment]
Fixed this error by:
1)	Accumulate the instruction operands type and value to the function hash.
2)	Accumulate the callee name to the caller hash.
```
   /// Accumulate the common parts of Call and Invoke instructions.
   template <typename T> void hashCallInvoke(const T *Instruction) {
-    FnHash.hashType(Instruction->getOperand(0)->getType());
+    // Accumulate the instruction operands type and value.
+    for (unsigned I = 0, E = Instruction->getNumOperands(); I != E; ++I) {
+      FnHash.hashType(Instruction->getOperand(I)->getType());
+      FnHash.hashValue(Instruction->getOperand(I));
+    }
     FnHash.hashNumber(Instruction->getCallingConv());
     FnHash.hashAttributeList(Instruction->getAttributes());
     hashOperandBundles(Instruction);
     FnHash.hashRangeMetadata(Instruction->getMetadata(LLVMContext::MD_range));
-    // Instruction operand 0 is the callee name. Accumulate it to the hash.
-    FnHash.hashValue(Instruction->getOperand(0));
     if (const Function *F = Instruction->getCalledFunction()) {
+      // Accumulate the callee name to the hash.
+      FnHash.hashMem(F->getName());
       if (!F->isDeclaration() && !F->hasAvailableExternallyLinkage())
         FnHash.getDependencies().emplace_back(F);
     }
```

**Comment by MaggieYingYi** [comment]
Fixed in commit 3102bc24e3.
